src/control/server/interceptors.go

```diff
@@ -205,3 +209,52 @@ func unaryStatusInterceptor(ctx context.Context, req interface{}, info *grpc.Una
 
 	return res, err
 }
+
+// isSentinelErr indicates whether or not the error is a sentinel
+// error used to convey a specific state to the client.
+func isSentinelErr(err error) bool {
+	return system.IsNotReady(err) || system.IsNotReplica(err) || system.IsNotLeader(err)
+}
+
+// unaryLoggingInterceptor generates a grpc.UnaryServerInterceptor that
+// will log an error if the RPC handler returned an error. If debugging is
+// enabled, it will also log the request and response messages.
+//
+// NB: This interceptor should be the last in the chain, i.e. first in the
+// list of interceptors passed to grpc.NewServer.
+func unaryLoggingInterceptor(log logging.Logger) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		startTime := time.Now()
+		res, err := handler(ctx, req)
+		elapsed := time.Since(startTime)
+		logErr := err
+		if logErr != nil {
+			// Unwrap the message if it's a gRPC status error.
+			if st, ok := status.FromError(err); ok {
+				logErr = proto.UnwrapError(st)
+			}
+		}
+
+		// If the error is nil or is not a sentinel error, log the request.
+		// We don't want to log sentinel errors because they're spammy and
+		// don't provide any useful debug information.
+		if logErr == nil || !isSentinelErr(logErr) {
+			if m, ok := req.(protoreflect.ProtoMessage); ok && log.EnabledFor(logging.LogLevelDebug) && proto.ShouldDebug(m) {
+				log.Debugf("gRPC request: %s", proto.Debug(m))
+			}
+		}
+
+		// Log the unwrapped error if it's not a sentinel error.
+		if logErr != nil {
+			if !isSentinelErr(logErr) {
+				log.Errorf("gRPC handler for %T failed: %s (elapsed: %s)", req, logErr, elapsed)
+			}
+			return res, err
+		}
```

Comment:
```suggestion
		} else {
		        // Log the unwrapped error if it's not a sentinel error.
			log.Errorf("gRPC handler for %T failed: %s (elapsed: %s)", req, logErr, elapsed)
		}
		if logErr != nil {
			return res, err
		}
```

Reply:
To me, it's clearer the other way. The intent is to avoid logging the request at all if it results in a sentinel error (e.g. ErrNotReplica), because the control client's leader/replica resolution mechanism will produce a lot of log spam otherwise. I expect that this will continue to evolve as we evaluate the logging going forward, though.